zebra/zebra_vxlan.c

```diff
@@ -183,6 +182,48 @@ static void zvni_deref_ip2mac(zebra_vni_t *zvni, zebra_mac_t *mac,
 			      int uninstall);
 
 /* Private functions */
+static int host_rb_entry_compare(const struct host_rb_entry *hle1,
+				   const struct host_rb_entry *hle2)
+{
+	if (hle1->p.family < hle2->p.family)
+		return -1;
+
+	if (hle1->p.family > hle2->p.family)
+		return 1;
+
+	if (hle1->p.prefixlen < hle2->p.prefixlen)
+		return -1;
+
+	if (hle1->p.prefixlen > hle2->p.prefixlen)
+		return 1;
+
+	if (hle1->p.family == AF_INET) {
+		if (hle1->p.u.prefix4.s_addr < hle2->p.u.prefix4.s_addr)
+			return -1;
+
+		if (hle1->p.u.prefix4.s_addr > hle2->p.u.prefix4.s_addr)
+			return 1;
+
+		return 0;
+	} else {
+		zlog_warn("%s: Unexpected family type: %d", __PRETTY_FUNCTION__,
+			  hle1->p.family);
+		return 0;
+	}
+}
+RB_GENERATE(host_rb_entry_rb, host_rb_entry, hl_entry,
+	    host_rb_entry_compare);
+
+static uint32_t rb_host_count(struct host_rb_entry_rb *hrbe)
+{
```

Comment:
Not sure if it's a good idea, but you could introduce a variable inside `zebra_mac_t_` to keep track of the number of hosts and update it whenever a new entry is added or removed to/from the tree. This would remove the need to loop throughout the whole tree whenever we need to count its number of elements. In the future we might also consider extending the rb-tree implementation to do this work for us (keeping track of the number of elements and provide an API to fetch this data).

Reply:
I'm not concerned about this, the counting of the rb entries appears to take very little time and it only happens in a couple of show commands.  I don't think there is a whole lot of need to add this functionality for this usage of the RB_Tree.